File: examples/verify-anchor.mjs
#!/usr/bin/env node
/**
 * Verify a published period report against its on-ledger anchor.
 *
 * The point of the anchor is not that arCCade can show you a report. It is
 * that a third party can prove a cycle was OMITTED from one. So this script
 * takes arCCade's published bytes as a SUSPECT rather than a source: it
 * recomputes every leaf from the rows, rebuilds the root, folds an inclusion
 * proof for each row, reassembles the anchor document field by field, and
 * re-derives the period totals — then compares each of those against what the
 * ledger was made to commit to. The report's own `leaves` and `merkleRoot`
 * are checked, never trusted.
 *
 * It also does the thing an example usually skips: it tampers with a row and
 * shows the same call refusing it. A verifier that only ever runs on good
 * input has not been tested, and neither has the reader's understanding of it.
 *
 *   node examples/verify-anchor.mjs              # the live TestNet report
 *   node examples/verify-anchor.mjs --offline    # the copy in fixtures/, no network
 *   node examples/verify-anchor.mjs --ledger     # anchors from your own participant
 *
 * Options:
 *   --source <url|dir>   where index.json and the reports live
 *                        (default https://audit.arccade.io/testnet)
 *   --anchors <file>     full anchor records  (default ./fixtures/anchors.json)
 *   --ledger             read VenuePeriodAnchor from LEDGER_URL/PARTY instead
 *   --offline            --source ./fixtures, and never touch the network
 *   --period <id>        check one period only, e.g. 2026-08-26
 *
 * Environment (only for --ledger):
 *   LEDGER_URL   JSON Ledger API base   (default http://localhost:7575)
 *   PARTY        a party that can see the anchor (the venue, operator or auditor)
 *   USER_ID      ledger API user        (default participant_admin)
 *   AUTH_TOKEN   bearer token, if your participant requires one
 *
 * Exit codes:
 *   0  everything reproduced
 *   1  a verification failed — a leaf, a root, a proof, an anchor or the chain
 *   2  everything reproduced EXCEPT the served bytes, which are not the bytes
 *      the anchor commits to. Separated because it is a different kind of
 *      finding: the rows are intact and the publisher's serialisation drifted.
 */

import { readFile } from 'node:fs/promises'
import { existsSync } from 'node:fs'
import { createHash } from 'node:crypto'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

import {
  canonInt,
  canonText,
  documentDigest,
  merkleEmpty,
  merkleProof,
  merkleRoot,
  periodLeaf,
  periodRowVerify,
} from '@arccade/game-sdk'

const HERE = dirname(fileURLToPath(import.meta.url))
const FIXTURES = join(HERE, 'fixtures')
const LIVE = 'https://audit.arccade.io/testnet'

const argv = process.argv.slice(2)
const flag = (name) => argv.includes(name)
const value = (name, fallback) => {
  const i = argv.indexOf(name)
  return i >= 0 && argv[i + 1] ? argv[i + 1] : fallback
}

const OFFLINE = flag('--offline')
const SOURCE = value('--source', OFFLINE ? FIXTURES : LIVE)
const ANCHORS = value('--anchors', join(FIXTURES, 'anchors.json'))
const FROM_LEDGER = flag('--ledger')
const ONLY = value('--period', null)

const LEDGER = (process.env.LEDGER_URL ?? 'http://localhost:7575').replace(/\/+$/, '')
const PARTY = process.env.PARTY
const USER_ID = process.env.USER_ID ?? 'participant_admin'
const TOKEN = process.env.AUTH_TOKEN ?? ''

const failures = []
const byteFindings = []
const fail = (what) => { failures.push(what); return false }

const sha256 = (bytes) => createHash('sha256').update(bytes).digest('hex')
const isUrl = (s) => /^https?:\/\//.test(s)
const step = (n, what) => console.log(`\n${n}. ${what}`)
const mark = (ok) => (ok ? 'ok  ' : 'FAIL')

/**
 * Reads one artifact and returns its EXACT BYTES.
 *
 * Not its parsed form and not a re-serialisation of it: `reportDigest` is a
 * hash of the file as served, so a verifier that parses first and hashes
 * second is checking a different object than the anchor committed to (T4).
 */
async function readBytes(base, name) {
  if (isUrl(base)) {
    if (OFFLINE) throw new Error(`--offline, but the source is a URL: ${base}`)
    const res = await fetch(`${base}/${name}`)
    if (!res.ok) throw new Error(`${base}/${name} -> ${res.status}`)
    return Buffer.from(await res.arrayBuffer())
  }
  return readFile(join(base, name))
}

/**
 * The anchor document, assembled from primitives.
 *
 * NO SHIPPED JAVASCRIPT CLIENT EXPORTS THIS. Daml computes it inside
 * `GameVenue_AnchorPeriod`; the Python client exports `anchor_document`; the
 * JavaScript and Java ones do not. So a verifier writes the fifteen fields
 * out, in this order, in schema `arccade.period-anchor` version 1 — which is
 * also the honest test of whether the scheme is really reimplementable from
 * the published description. Values arrive from JSON as strings; `canonInt`
 * takes them through BigInt, so no decimal ever appears in the bytes.
 */
function anchorDigestOf(a) {
  return documentDigest('arccade.period-anchor', 1, [
    ['venueId', canonText(a.venueId)],
    ['periodId', canonText(a.periodId)],
    ['periodStartMicros', canonInt(a.periodStartMicros)],
    ['periodEndMicros', canonInt(a.periodEndMicros)],
    ['cycleCount', canonInt(a.cycleCount)],
    ['committedUnits', canonInt(a.committedUnits)],
    ['feeUnits', canonInt(a.feeUnits)],
    ['returnedUnits', canonInt(a.returnedUnits)],
    ['forfeitedUnits', canonInt(a.forfeitedUnits)],
    ['payoutUnits', canonInt(a.payoutUnits)],
    ['qualifyingTxCount', canonInt(a.qualifyingTxCount)],
    ['nonQualifyingTxCount', canonInt(a.nonQualifyingTxCount)],
    ['merkleRootHex', canonText(a.merkleRootHex)],
    ['reportDigest', canonText(a.reportDigest)],
    // Empty on the first period of a chain, and that emptiness is signed.
    ['prevAnchorDigest', canonText(a.prevAnchorDigest)],
  ])
}

/**
 * Period totals DERIVED FROM THE ROWS.
 *
 * A correct Merkle root says nothing whatever about the summary fields — the
 * root commits to the rows, not to the arithmetic over them. Recomputing the
 * totals is what stops a venue from publishing a verifiable root beside a
 * fabricated headline number.
 */
function totalsOf(rows) {
  const sum = (f) => rows.reduce((acc, r) => acc + BigInt(r[f]), 0n)
  return {
    cycleCount: BigInt(rows.length),
    committedUnits: sum('committedUnits'),
    feeUnits: sum('feeUnits'),
    returnedUnits: sum('returnedUnits'),
    forfeitedUnits: sum('forfeitedUnits'),
    payoutUnits: sum('payoutUnits'),
  }
}

/** Anchors from a participant's ACS — the strongest source there is. */
async function anchorsFromLedger() {
  if (!PARTY) throw new Error('--ledger needs PARTY: a party that can see the anchor.')
  const call = async (path, body) => {
    const res = await fetch(LEDGER + path, {
      method: body ? 'POST' : 'GET',
      headers: {
        'Content-Type': 'application/json',
        ...(TOKEN ? { Authorization: `Bearer ${TOKEN}` } : {}),
      },
      body: body ? JSON.stringify(body) : undefined,
    })
    const text = await res.text()
    if (!res.ok) throw new Error(`${path} -> ${res.status}\n${text.slice(0, 400)}`)
    return JSON.parse(text)
  }
  const { offset } = await call('/v2/state/ledger-end')
  const acs = await call('/v2/state/active-contracts', {
    filter: { filtersByParty: { [PARTY]: { cumulative: [
      { identifierFilter: { WildcardFilter: { value: { includeCreatedEventBlob: false } } } },
    ] } } },
    verbose: false,
    activeAtOffset: offset,
    userId: USER_ID,
  })
  return acs
    .map((e) => e?.contractEntry?.JsActiveContract?.createdEvent)
    .filter((c) => c && (c.templateId ?? '').includes('ArCCade.GameSdk.Audit:VenuePeriodAnchor'))
    .map((c) => ({ contractId: c.contractId, createArgument: c.createArgument }))
}

// ---------------------------------------------------------------------------

console.log(`source  ${SOURCE}${OFFLINE ? '   (offline)' : ''}`)

step(1, 'The index, and where the anchors are coming from')
const index = JSON.parse((await readBytes(SOURCE, 'index.json')).toString('utf8'))
console.log(`   ${index.count} period(s) published`)

let anchors = []
let anchorSource = 'none'
if (FROM_LEDGER) {
  anchors = await anchorsFromLedger()
  anchorSource = `${LEDGER} (ACS)`
} else if (existsSync(ANCHORS)) {
  anchors = JSON.parse(await readFile(ANCHORS, 'utf8')).anchors ?? []
  anchorSource = ANCHORS
}
console.log(`   anchors from ${anchorSource}: ${anchors.length}`)
if (anchors.length === 0) {
  // index.json carries the anchored root and digest, so the leaf, root and
  // proof checks still run — but the anchor DOCUMENT cannot be reassembled
  // from a summary, so step 6 will be skipped rather than quietly passed.
  console.log('   the anchor document check needs the contract itself; use --ledger')
}
const anchorFor = (periodId) =>
  anchors.map((a) => a.createArgument).find((a) => a.periodId === periodId) ?? null

const reports = index.reports.filter((r) => !ONLY || r.periodId === ONLY)
if (reports.length === 0) throw new Error(`no published period matches --period ${ONLY}`)

let previousAnchorDigest = null

for (const meta of reports) {
  // The step counter restarts under each period heading: every period is
  // checked the same way, and a reader comparing two of them should be
  // comparing step 4 with step 4.
  let n = 0
  console.log(`\n${'─'.repeat(72)}`)
  console.log(`${meta.venueId}  period ${meta.periodId}  (${meta.cycleCount} cycles)`)
  const anchor = anchorFor(meta.periodId)

  step(++n, 'The report, hashed as bytes')
  const bytes = await readBytes(SOURCE, meta.name)
  const served = sha256(bytes)
  const report = JSON.parse(bytes.toString('utf8'))
  console.log(`   ${meta.name}`)
  console.log(`   sha256 ${served}`)
  // Reproducible with no library at all — which is the claim being made:
  //   sha256sum tradewars_testnet-arena-v2_2026-08-27.json
  console.log(`   ${mark(served === meta.servedDigest)}  matches index.servedDigest`)
  if (served !== meta.servedDigest) fail(`${meta.periodId}: served bytes differ from index.servedDigest`)

  step(++n, 'Every leaf, recomputed from its row')
  // periodLeaf renders the row as `arccade.cycle-audit-row` v1 and hashes it.
  // Amounts are integer 1e-10 units and times are integer microseconds — the
  // report gives both as strings precisely so no JSON number ever rounds one.
  const leaves = report.rows.map(periodLeaf)
  let leavesOk = leaves.length === (report.leaves ?? []).length
  leaves.forEach((leaf, i) => { if (leaf !== report.leaves?.[i]) leavesOk = false })
  for (const [i, leaf] of leaves.entries()) {
    const same = leaf === report.leaves?.[i]
    console.log(`   ${mark(same)}  [${i}] ${report.rows[i].cycleId}  ${leaf.slice(0, 16)}…`)
  }
  if (leaves.length === 0) console.log('   (no rows — an empty period is still anchored)')
  if (!leavesOk) fail(`${meta.periodId}: recomputed leaves differ from the published ones`)

  step(++n, 'The root, rebuilt from the leaves')
  // A lone trailing node is PROMOTED unchanged rather than duplicated: the
  // Bitcoin convention (CVE-2012-2459) lets [a,b,c] and [a,b,c,c] produce one
  // root, and this tree refuses that by construction.
  const root = merkleRoot(leaves)
  console.log(`   ${root}${leaves.length === 0 ? '   (= merkleEmpty)' : ''}`)
  if (leaves.length === 0 && root !== merkleEmpty()) fail(`${meta.periodId}: empty root is wrong`)
  console.log(`   ${mark(root === report.merkleRoot)}  matches the report's merkleRoot`)
  console.log(`   ${mark(root === meta.anchoredRoot)}  matches the ANCHORED root`)
  if (root !== report.merkleRoot) fail(`${meta.periodId}: rebuilt root differs from the report`)
  if (root !== meta.anchoredRoot) fail(`${meta.periodId}: rebuilt root differs from the anchor`)

  step(++n, 'An inclusion proof for every row')
  if (leaves.length === 0) {
    // merkleProof returns [] for any index into an empty period, and folding
    // [] returns the leaf unchanged. An empty proof must never be read as
    // proof of anything, so there is nothing to run here and saying so is the
    // honest output.
    console.log('   nothing to prove: no rows. An empty proof proves nothing.')
  } else {
    for (const [i, row] of report.rows.entries()) {
      const proof = merkleProof(i, leaves)
      // periodRowVerify, not merkleVerify: folding a bare hash returns true
      // for an internal node too, because the fold cannot know what it
      // started from. Deriving the leaf FROM THE ROW is what binds the claim
      // "this is a cycle" to the row schema.
      const ok = periodRowVerify(row, proof, root)
      console.log(`   ${mark(ok)}  [${i}] ${row.cycleId}  ${proof.length} step(s)`)
      if (!ok) fail(`${meta.periodId}: inclusion proof failed for ${row.cycleId}`)
    }

    // The check that matters, and the one an example usually leaves out.
    const tampered = { ...report.rows[0], returnedUnits: String(BigInt(report.rows[0].returnedUnits) - 1n) }
    const refused = periodRowVerify(tampered, merkleProof(0, leaves), root) === false
    console.log(`   ${mark(refused)}  the same proof REFUSES the row with returnedUnits-1`)
    console.log(`         ${report.rows[0].returnedUnits} -> ${tampered.returnedUnits}`)
    if (!refused) fail(`${meta.periodId}: a tampered row verified against the root`)
  }

  step(++n, 'Totals, re-derived from the rows')
  const totals = totalsOf(report.rows)
  if (anchor) {
    for (const [field, got] of Object.entries(totals)) {
      const same = got === BigInt(anchor[field])
      console.log(`   ${mark(same)}  ${field.padEnd(16)} ${got}`)
      if (!same) fail(`${meta.periodId}: ${field} disagrees with the anchor (${anchor[field]})`)
    }
  } else {
    console.log(`   cycleCount ${totals.cycleCount} vs index ${meta.cycleCount}`)
    if (totals.cycleCount !== BigInt(meta.cycleCount)) {
      fail(`${meta.periodId}: row count disagrees with the index`)
    }
    console.log('   the rest need the anchor contract; use --ledger or fixtures/anchors.json')
  }

  step(++n, 'The anchor document, reassembled field by field')
  if (!anchor) {
    console.log('   SKIPPED — no anchor contract available for this period')
  } else {
    const digest = anchorDigestOf(anchor)
    console.log(`   ${digest}`)
    console.log(`   ${mark(digest === anchor.anchorDigest)}  matches anchorDigest on the contract`)
    console.log(`   ${mark(anchor.merkleRootHex === root)}  the anchor commits to the root we rebuilt`)
    if (digest !== anchor.anchorDigest) fail(`${meta.periodId}: anchor document does not reproduce`)
    if (anchor.merkleRootHex !== root) fail(`${meta.periodId}: anchor root differs from the rebuilt root`)
    // reportUri is a field of the CONTRACT, not of the document: the place a
    // report is served from is not part of the commitment, its bytes are.
    console.log(`   reportUri (not covered by the digest): ${anchor.reportUri}`)

    step(++n, 'The chain')
    if (previousAnchorDigest === null) {
      const first = anchor.prevAnchorDigest === ''
      console.log(`   prevAnchorDigest ${anchor.prevAnchorDigest === '' ? '""  (start of the chain)' : anchor.prevAnchorDigest}`)
      if (!first) console.log('   this is not the first period — run without --period to walk it')
    } else {
      const linked = anchor.prevAnchorDigest === previousAnchorDigest
      console.log(`   ${mark(linked)}  prevAnchorDigest is the previous period's anchorDigest`)
      console.log(`         ${anchor.prevAnchorDigest}`)
      // A missing period is a broken link, not an absence you have to notice.
      if (!linked) fail(`${meta.periodId}: chain link broken — a period is missing or reordered`)
    }
    previousAnchorDigest = anchor.anchorDigest
  }

  step(++n, 'The bytes the anchor actually commits to')
  const anchoredDigest = anchor?.reportDigest ?? meta.anchoredDigest
  const bytesMatch = served === anchoredDigest
  console.log(`   served   ${served}`)
  console.log(`   anchored ${anchoredDigest}`)
  console.log(`   ${mark(bytesMatch)}  the file served is the file anchored`)
  if (!bytesMatch) {
    byteFindings.push(meta.periodId)
    console.log(`
   T4. The rows are intact — every leaf and the root reproduced above — but
   these bytes are not the bytes the ledger was made to commit to. Publish a
   report's bytes once and serve them byte-stable forever; re-serialising to
   pretty-print it is a breaking change. index.json records both digests, so
   this was noticed rather than hidden.`)
  }
}

// ---------------------------------------------------------------------------

console.log(`\n${'─'.repeat(72)}`)
if (failures.length > 0) {
  console.error(`\n${failures.length} verification(s) failed:`)
  for (const f of failures) console.error(`  - ${f}`)
  process.exit(1)
}
if (byteFindings.length > 0) {
  console.log(`
Rows, roots, proofs, anchors and the chain all reproduce.

The served bytes do not match the anchored bytes for: ${byteFindings.join(', ')}.
Exit 2 — see T4 in docs/INTEGRATION.md. This is a finding about how the report
was published, not about what happened on the ledger.
`)
  process.exit(2)
}
console.log(`
Everything reproduced, from the bytes up, with nothing taken from arCCade but
the report itself. That is the whole claim: an omitted cycle cannot survive
this check, because the root the ledger holds would no longer be the root the
rows produce.
`)

File: examples/verify_anchor.py
#!/usr/bin/env python3
"""Verify a published period report against its on-ledger anchor — in Python.

The JavaScript verify-anchor.mjs and this file check the same artifacts and
must print the same hexadecimal. That is the point of having both: the scheme
is only independently verifiable if a second implementation, reading nothing
but the published description, lands on the same bytes. Where they differ, the
difference is what is worth reading.

  * Python's client DOES export ``anchor_document``; JavaScript's and Java's do
    not, so the .mjs assembles the fifteen fields by hand. Here the SDK does it.
  * ``canon_int`` refuses ``bool`` and refuses a decimal string. JavaScript's
    ``BigInt`` would happily turn ``true`` into ``1``, so the same caller
    mistake would produce two different documents that both looked successful.
  * Report order (T14) breaks ties differently in the two languages. It does
    not bite here — no two rows share a ``committedAtMicros`` — but a verifier
    that RE-SORTS the rows rather than reading them in file order is choosing a
    collation, and the two languages do not choose the same one. So this reads
    them in file order and checks the root, which is the thing that must match.

  python3 examples/verify_anchor.py             # the live TestNet report
  python3 examples/verify_anchor.py --offline   # the copy in fixtures/, no network

Options and exit codes are the same as verify-anchor.mjs:
  0  everything reproduced
  1  a verification failed
  2  reproduced, except that the served bytes are not the anchored bytes (T4)
"""

from __future__ import annotations

import argparse
import hashlib
import json
import sys
import urllib.request
from pathlib import Path

HERE = Path(__file__).resolve().parent
FIXTURES = HERE / "fixtures"
LIVE = "https://audit.arccade.io/testnet"

# Run from a clone with no install: the sibling source tree is the same code
# the published package ships. `pip install arccade-game-sdk` also works and
# takes precedence, because this only appends.
sys.path.append(str(HERE.parent / "python"))

from arccade_game_sdk import (  # noqa: E402
    anchor_digest,
    merkle_empty,
    merkle_proof,
    merkle_root,
    period_leaf,
    period_row_verify,
)

failures: list[str] = []
byte_findings: list[str] = []


def fail(what: str) -> None:
    failures.append(what)


def mark(ok: bool) -> str:
    return "ok  " if ok else "FAIL"


def step(n: int, what: str) -> int:
    print(f"\n{n}. {what}")
    return n


def read_bytes(base: str, name: str, offline: bool) -> bytes:
    """The EXACT bytes, never a re-serialisation.

    ``reportDigest`` commits to the file as served. Parsing first and hashing
    second would check a different object than the anchor committed to (T4).
    """
    if base.startswith(("http://", "https://")):
        if offline:
            raise SystemExit(f"--offline, but the source is a URL: {base}")
        with urllib.request.urlopen(f"{base}/{name}", timeout=30) as r:
            return r.read()
    return (Path(base) / name).read_bytes()


def totals_of(rows: list[dict]) -> dict[str, int]:
    """Totals DERIVED FROM THE ROWS.

    A correct root says nothing about the summary fields: it commits to the
    rows, not to the arithmetic over them.
    """
    def s(field: str) -> int:
        return sum(int(r[field]) for r in rows)

    return {
        "cycleCount": len(rows),
        "committedUnits": s("committedUnits"),
        "feeUnits": s("feeUnits"),
        "returnedUnits": s("returnedUnits"),
        "forfeitedUnits": s("forfeitedUnits"),
        "payoutUnits": s("payoutUnits"),
    }


def main() -> int:
    ap = argparse.ArgumentParser(add_help=True)
    ap.add_argument("--source", default=None, help="url or directory holding index.json")
    ap.add_argument("--anchors", default=str(FIXTURES / "anchors.json"))
    ap.add_argument("--offline", action="store_true")
    ap.add_argument("--period", default=None)
    args = ap.parse_args()

    source = args.source or (str(FIXTURES) if args.offline else LIVE)
    print(f"source  {source}{'   (offline)' if args.offline else ''}")

    step(1, "The index, and where the anchors are coming from")
    index = json.loads(read_bytes(source, "index.json", args.offline))
    print(f"   {index['count']} period(s) published")

    anchors: list[dict] = []
    anchor_path = Path(args.anchors)
    if anchor_path.exists():
        anchors = [a["createArgument"] for a in json.loads(anchor_path.read_text())["anchors"]]
    print(f"   anchors from {anchor_path if anchors else 'nowhere'}: {len(anchors)}")

    reports = [r for r in index["reports"] if not args.period or r["periodId"] == args.period]
    if not reports:
        raise SystemExit(f"no published period matches --period {args.period}")

    previous_anchor_digest: str | None = None

    for meta in reports:
        anchor = next((a for a in anchors if a["periodId"] == meta["periodId"]), None)
        print("\n" + "─" * 72)
        print(f"{meta['venueId']}  period {meta['periodId']}  ({meta['cycleCount']} cycles)")
        n = 0

        n = step(n + 1, "The report, hashed as bytes")
        raw = read_bytes(source, meta["name"], args.offline)
        served = hashlib.sha256(raw).hexdigest()
        report = json.loads(raw)
        print(f"   {meta['name']}")
        print(f"   sha256 {served}")
        print(f"   {mark(served == meta['servedDigest'])}  matches index.servedDigest")
        if served != meta["servedDigest"]:
            fail(f"{meta['periodId']}: served bytes differ from index.servedDigest")

        n = step(n + 1, "Every leaf, recomputed from its row")
        leaves = [period_leaf(row) for row in report["rows"]]
        for i, leaf in enumerate(leaves):
            same = leaf == report["leaves"][i]
            print(f"   {mark(same)}  [{i}] {report['rows'][i]['cycleId']}  {leaf[:16]}…")
            if not same:
                fail(f"{meta['periodId']}: leaf {i} does not reproduce")
        if not leaves:
            print("   (no rows — an empty period is still anchored)")
        if len(leaves) != len(report["leaves"]):
            fail(f"{meta['periodId']}: leaf count differs from the published one")

        n = step(n + 1, "The root, rebuilt from the leaves")
        # A lone trailing node is PROMOTED, not duplicated: the Bitcoin
        # convention (CVE-2012-2459) lets [a,b,c] and [a,b,c,c] share a root.
        root = merkle_root(leaves)
        print(f"   {root}{'   (= merkle_empty)' if not leaves else ''}")
        if not leaves and root != merkle_empty():
            fail(f"{meta['periodId']}: empty root is wrong")
        print(f"   {mark(root == report['merkleRoot'])}  matches the report's merkleRoot")
        print(f"   {mark(root == meta['anchoredRoot'])}  matches the ANCHORED root")
        if root != report["merkleRoot"]:
            fail(f"{meta['periodId']}: rebuilt root differs from the report")
        if root != meta["anchoredRoot"]:
            fail(f"{meta['periodId']}: rebuilt root differs from the anchor")

        n = step(n + 1, "An inclusion proof for every row")
        if not leaves:
            # merkle_proof returns [] for any index into an empty period, and
            # folding [] returns the leaf unchanged. An empty proof must never
            # be read as proof of anything.
            print("   nothing to prove: no rows. An empty proof proves nothing.")
        else:
            for i, row in enumerate(report["rows"]):
                proof = merkle_proof(i, leaves)
                # period_row_verify, not merkle_verify: folding a bare hash
                # returns True for an internal node too. Deriving the leaf from
                # the row binds the claim "this is a cycle" to the row schema.
                ok = period_row_verify(row, proof, root)
                print(f"   {mark(ok)}  [{i}] {row['cycleId']}  {len(proof)} step(s)")
                if not ok:
                    fail(f"{meta['periodId']}: inclusion proof failed for {row['cycleId']}")

            tampered = dict(report["rows"][0])
            tampered["returnedUnits"] = str(int(tampered["returnedUnits"]) - 1)
            refused = period_row_verify(tampered, merkle_proof(0, leaves), root) is False
            print(f"   {mark(refused)}  the same proof REFUSES the row with returnedUnits-1")
            print(f"         {report['rows'][0]['returnedUnits']} -> {tampered['returnedUnits']}")
            if not refused:
                fail(f"{meta['periodId']}: a tampered row verified against the root")

        n = step(n + 1, "Totals, re-derived from the rows")
        totals = totals_of(report["rows"])
        if anchor:
            for field, got in totals.items():
                same = got == int(anchor[field])
                print(f"   {mark(same)}  {field:<16} {got}")
                if not same:
                    fail(f"{meta['periodId']}: {field} disagrees with the anchor")
        else:
            print(f"   cycleCount {totals['cycleCount']} vs index {meta['cycleCount']}")
            print("   the rest need the anchor contract; see fixtures/anchors.json")

        n = step(n + 1, "The anchor document, reassembled field by field")
        if not anchor:
            print("   SKIPPED — no anchor contract available for this period")
        else:
            # The SDK's own builder, unlike the JavaScript side. Same fifteen
            # fields, same order, same bytes — that equality is the claim.
            digest = anchor_digest(anchor)
            print(f"   {digest}")
            print(f"   {mark(digest == anchor['anchorDigest'])}  matches anchorDigest on the contract")
            print(f"   {mark(anchor['merkleRootHex'] == root)}  the anchor commits to the root we rebuilt")
            if digest != anchor["anchorDigest"]:
                fail(f"{meta['periodId']}: anchor document does not reproduce")
            if anchor["merkleRootHex"] != root:
                fail(f"{meta['periodId']}: anchor root differs from the rebuilt root")
            # reportUri is a field of the CONTRACT, not of the document: where
            # a report is served from is not part of the commitment.
            print(f"   reportUri (not covered by the digest): {anchor['reportUri']}")

            n = step(n + 1, "The chain")
            if previous_anchor_digest is None:
                shown = '""  (start of the chain)' if anchor["prevAnchorDigest"] == "" else anchor["prevAnchorDigest"]
                print(f"   prevAnchorDigest {shown}")
                if anchor["prevAnchorDigest"] != "":
                    print("   this is not the first period — run without --period to walk it")
            else:
                linked = anchor["prevAnchorDigest"] == previous_anchor_digest
                print(f"   {mark(linked)}  prevAnchorDigest is the previous period's anchorDigest")
                print(f"         {anchor['prevAnchorDigest']}")
                if not linked:
                    fail(f"{meta['periodId']}: chain link broken — a period is missing or reordered")
            previous_anchor_digest = anchor["anchorDigest"]

        n = step(n + 1, "The bytes the anchor actually commits to")
        anchored = anchor["reportDigest"] if anchor else meta["anchoredDigest"]
        print(f"   served   {served}")
        print(f"   anchored {anchored}")
        print(f"   {mark(served == anchored)}  the file served is the file anchored")
        if served != anchored:
            byte_findings.append(meta["periodId"])
            print(
                "\n   T4. The rows are intact — every leaf and the root reproduced above —\n"
                "   but these bytes are not the bytes the ledger was made to commit to.\n"
                "   Publish a report's bytes once and serve them byte-stable forever."
            )

    print("\n" + "─" * 72)
    if failures:
        print(f"\n{len(failures)} verification(s) failed:", file=sys.stderr)
        for f in failures:
            print(f"  - {f}", file=sys.stderr)
        return 1
    if byte_findings:
        print(
            f"\nRows, roots, proofs, anchors and the chain all reproduce.\n\n"
            f"The served bytes do not match the anchored bytes for: {', '.join(byte_findings)}.\n"
            f"Exit 2 — see T4 in docs/INTEGRATION.md.\n"
        )
        return 2
    print(
        "\nEverything reproduced, from the bytes up, and the hexadecimal above is\n"
        "identical to what verify-anchor.mjs prints. Two implementations, one\n"
        "canonical form — which is what makes the anchor evidence rather than a\n"
        "record arCCade keeps.\n"
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
